### PR TITLE
WIP: Add v1alpha3 support

### PR DESCRIPTION
**Pull request by jichenjc**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by jichenjc** [comment]
encountered some issue...  ask here
https://github.com/kubernetes-sigs/cluster-api/issues/2219

**Comment by ncdc** [comment]
I have a pretty big review backlog at this time, but I'll try to make some time for it.

@CecileRobertMichon @juan-lee @rudoi @wfernandes @yastij @noamran @chuckha @randomvariable @detiber @justaugustus if any of you has time to help review for v1alpha3 changes...

**Comment by wfernandes** [comment]
I can make some time to take a look at this.

**Comment by randomvariable** [comment]
I'll have a look tomorrow afternoon (GMT).

**Comment by wfernandes** [comment]
@jichenjc Apologies for not taking a look at this earlier. Is this still WIP?

**Comment by jichenjc** [comment]
@wfernandes 
yes, this is still WIP because I have some trouble in creating the cluster until now with v1alpha3
so I might need more actions for this PR... thanks for the reminder

**Comment by sbueringer** [comment]
@jichenjc any news?

**Comment by jichenjc** [comment]
latest status is I can create master but not worker node because of some unknown issue
too busy this week and I will make sure make some progress on this next week

File: config/crd/bases/infrastructure.cluster.x-k8s.io_openstackclusters.yaml [hunk 1]
```diff
@@ -13,481 +13,977 @@ spec:
...
-                  type: string
-                limit:
+              ready:
+                type: boolean
+            required:
+            - ready
+            type: object
+        type: object
+    served: true
+    storage: true
```
Comment (hidekazuna) [inline config/crd/bases/infrastructure.cluster.x-k8s.io_openstackclusters.yaml]: storage of v1alpha2 should be false. I got the following error.
```
Error from server (Invalid): error when creating "infrastructure-components.yaml": CustomResourceDefinition.apiextensions.k8s.io "openstackclusters.infrastructure.cluster.x-k8s.io" is invalid: [spec.versions: Invalid value: []apiextensions.CustomResourceDefinitionVersion{apiextensions.CustomResourceDefinitionVersion{Name:"v1alpha2", Served:true, Storage:true, Schema:(*apiextensions.CustomResourceValidation)(0xc00ba6d1d0), Subresources:(*apiextensions.CustomResourceSubresources)(nil), AdditionalPrinterColumns:[]apiextensions.CustomResourceColumnDefinition(nil)}, apiextensions.CustomResourceDefinitionVersion{Name:"v1alpha3", Served:true, Storage:true, Schema:(*apiextensions.CustomResourceValidation)(0xc00ba6d1d8), Subresources:(*apiextensions.CustomResourceSubresources)(nil), AdditionalPrinterColumns:[]apiextensions.CustomResourceColumnDefinition(nil)}}: must have exactly one version marked as storage version, status.storedVersions: Invalid value: apiextensions.CustomResourceDefinitionVersion{Name:"v1alpha3", Served:true, Storage:true, Schema:(*apiextensions.CustomResourceValidation)(0xc00ba6d1d8), Subresources:(*apiextensions.CustomResourceSubresources)(nil), AdditionalPrinterColumns:[]apiextensions.CustomResourceColumnDefinition(nil)}: must have the storage version v1alpha3]
```

File: config/crd/bases/infrastructure.cluster.x-k8s.io_openstackmachines.yaml [hunk 2]
```diff
@@ -320,6 +320,9 @@ spec:
   - name: v1alpha2
     served: true
     storage: true
```
Comment (hidekazuna) [inline config/crd/bases/infrastructure.cluster.x-k8s.io_openstackmachines.yaml]: storage of v1alpha2 should be false. I got the following error.
```
Error from server (Invalid): error when creating "infrastructure-components.yaml": CustomResourceDefinition.apiextensions.k8s.io "openstackmachines.infrastructure.cluster.x-k8s.io" is invalid: [spec.versions: Invalid value: []apiextensions.CustomResourceDefinitionVersion{apiextensions.CustomResourceDefinitionVersion{Name:"v1alpha2", Served:true, Storage:true, Schema:(*apiextensions.CustomResourceValidation)(nil), Subresources:(*apiextensions.CustomResourceSubresources)(nil), AdditionalPrinterColumns:[]apiextensions.CustomResourceColumnDefinition(nil)}, apiextensions.CustomResourceDefinitionVersion{Name:"v1alpha3", Served:true, Storage:true, Schema:(*apiextensions.CustomResourceValidation)(nil), Subresources:(*apiextensions.CustomResourceSubresources)(nil), AdditionalPrinterColumns:[]apiextensions.CustomResourceColumnDefinition(nil)}}: must have exactly one version marked as storage version, status.storedVersions: Invalid value: apiextensions.CustomResourceDefinitionVersion{Name:"v1alpha3", Served:true, Storage:true, Schema:(*apiextensions.CustomResourceValidation)(nil), Subresources:(*apiextensions.CustomResourceSubresources)(nil), AdditionalPrinterColumns:[]apiextensions.CustomResourceColumnDefinition(nil)}: must have the storage version v1alpha3]
```

File: config/crd/bases/infrastructure.cluster.x-k8s.io_openstackmachinetemplates.yaml [hunk 3]
```diff
@@ -293,6 +293,9 @@ spec:
   - name: v1alpha2
     served: true
     storage: true
```
Comment (hidekazuna) [inline config/crd/bases/infrastructure.cluster.x-k8s.io_openstackmachinetemplates.yaml]: storage of v1alpha2 should be false. I got the following error.
```
Error from server (Invalid): error when creating "infrastructure-components.yaml": CustomResourceDefinition.apiextensions.k8s.io "openstackmachinetemplates.infrastructure.cluster.x-k8s.io" is invalid: [spec.versions: Invalid value: []apiextensions.CustomResourceDefinitionVersion{apiextensions.CustomResourceDefinitionVersion{Name:"v1alpha2", Served:true, Storage:true, Schema:(*apiextensions.CustomResourceValidation)(nil), Subresources:(*apiextensions.CustomResourceSubresources)(nil), AdditionalPrinterColumns:[]apiextensions.CustomResourceColumnDefinition(nil)}, apiextensions.CustomResourceDefinitionVersion{Name:"v1alpha3", Served:true, Storage:true, Schema:(*apiextensions.CustomResourceValidation)(nil), Subresources:(*apiextensions.CustomResourceSubresources)(nil), AdditionalPrinterColumns:[]apiextensions.CustomResourceColumnDefinition(nil)}}: must have exactly one version marked as storage version, status.storedVersions: Invalid value: apiextensions.CustomResourceDefinitionVersion{Name:"v1alpha3", Served:true, Storage:true, Schema:(*apiextensions.CustomResourceValidation)(nil), Subresources:(*apiextensions.CustomResourceSubresources)(nil), AdditionalPrinterColumns:[]apiextensions.CustomResourceColumnDefinition(nil)}: must have the storage version v1alpha3]
```

**Comment by ncdc** [comment]
CRD v1 is required for Cluster API v1alpha3. Management clusters must be at least Kubernetes v1.16. Our changes doc (https://master.cluster-api.sigs.k8s.io/developer/providers/v1alpha2-to-v1alpha3.html) mentions CRD v1, although it's a bit hidden in the multi-tenancy section.

@vincepri I think we should make the changes doc clearer to indicate 1.16 and CRDv1 are required - WDYT?

**Comment by vincepri** [comment]
@ncdc sounds good to me, I'll start working on it

File: config/crd/bases/infrastructure.cluster.x-k8s.io_openstackclusters.yaml (hunk 1 above)
Comment (jichenjc) [reply config/crd/bases/infrastructure.cluster.x-k8s.io_openstackclusters.yaml]: the problem is the conflict when create alpha2 and alpha3 ,I think it's something not related to storage?

**Comment by jichenjc** [comment]
thanks for the info , I will switch my env to 1.16 and update CRD as well

**Comment by sbueringer** [comment]
@jichenjc is there any way I can help with this PR?

**Comment by sbueringer** [comment]
@jichenjc I have a working version. Including notes. I worked through the whole migration guide based on this PR

**Comment by sbueringer** [comment]
@vincepri @ncdc @detiber 

The migration guide for v1alpha3 is really very good. Thx for that! :)

I saw some minor inconsistencies/ issues:
* the following seeems redundant:
    * manager_label_patch.yaml to config/manager
    * manager_label_patch.yaml to trash.
* CAPA is not using the manager_role_aggregation_patch.yaml. Our controller didn't work without
* CAPA is not using manager_prometheus_metrics_patch.yaml
* Occurrences ./config/default/manager_pull_policy.yaml must also be changed to ./config/manager/manager_pull_policy.yaml

But maybe I'm interpreting it incorrectly

**Comment by jichenjc** [comment]
let's switch to @sbueringer 's PR ,as I again buried by some other things recently

**Pull request closed without merging**